### PR TITLE
Support push

### LINKED ISSUE #34: support push

### PR DESCRIPTION
**Pull request by overvenus**
Resolved #34 

PTAL @siddontang @disksing

### CODE REVIEW

File: src/push.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    let max_len = 256;
+    let mut name = vec![0u8; max_len];
+    match unsafe {
+        gethostname(name.as_mut_ptr() as *mut libc::c_char,
+                    max_len as libc::size_t)
+    } {
+        0 => {
+            let last_char = name.iter().position(|byte| *byte == 0).unwrap_or(max_len);
+            labels!{
+                "instance".to_owned() => str::from_utf8(&name[..last_char]).unwrap().to_owned(),
```
Comment (siddontang) [inline src/push.rs]: go client ignores the error, I think we should check it.

File: src/push.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+/// nor any of the grouping labels.
+///
+/// You can use just host:port or ip:port as url, in which case 'http://' is
+/// added automatically. You can also include the schema in the URL. However, do
+/// not include the '/metrics/jobs/...' part.
+///
+/// Note that all previously pushed metrics with the same job and other grouping
+/// labels will be replaced with the metrics pushed by this call. (It uses HTTP
+/// method 'PUT' to push to the Pushgateway.)
+pub fn push_from_gather(job: &str,
```
Comment (siddontang) [inline src/push.rs]: support push from collector too.

File: src/push.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        if lv.contains('/') {
+            return Err(Error::Msg(format!("value of grouping label {} contains '/': {}", ln, lv)));
+        }
+        url_components.push(ln.to_owned());
+        url_components.push(lv.to_owned());
+    }
+
+    push_url = format!("{}/metrics/job/{}", push_url, url_components.join("/"));
+
+    let encoder = TextEncoder::new();
```
Comment (siddontang) [inline src/push.rs]: push uses text format?

File: src/push.rs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        grouping: HashMap<String, String>,
+        url: &str,
+        mfs: Vec<proto::MetricFamily>,
+        method: &str)
+        -> Result<()> {
+
+    let mut push_url = if url.contains("://") {
+        url.to_owned()
+    } else {
+        "http://".to_owned() + url
```
Comment (siddontang) [inline src/push.rs]: use format

File: src/push.rs [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+use hyper::header::ContentType;
+
+use proto;
+use errors::{Result, Error};
+use encoder::{Encoder, TextEncoder};
+
+lazy_static!{
+    static ref HTTP_CLIENT: Client = Client::with_pool_config(
+            Config{
+                max_idle: 1,
```
Comment (siddontang) [inline src/push.rs]: use a const var

File: src/push.rs (hunk 3 above)
Comment (overvenus) [inline src/push.rs]: For now, we only support text format :-(
But it works.

File: examples/example_push.rs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    static ref PUSH_REQ_HISTOGRAM: Histogram = register_histogram!(
+        histogram_opts!(
+            "example_push_request_duration_seconds",
+            "The push request latencies in seconds."
+        )
+    ).unwrap();
+}
+
+fn main() {
+    println!("Pushing, please wait 10 seconds ...");
```
Comment (siddontang) [inline examples/example_push.rs]: we must tell user to start prometheus server first.
Add a flag for the prometheus server, not fixed `127.0.0.1:9091`

File: examples/example_push.rs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    for _ in 0..5 {
+        thread::sleep(time::Duration::from_secs(2));
+        PUSH_COUNTER.inc();
+        let metric_familys = prometheus::gather();
+        let _timer = PUSH_REQ_HISTOGRAM.start_timer(); // drop as observe
+        prometheus::push_metrics("example_push",
+                                 labels!{"instance".to_owned() => "HAL-9000".to_owned(),},
```
Comment (siddontang) [inline examples/example_push.rs]: what is `HAL-9000`?

File: examples/example_push.rs (hunk 6 above)
Comment (overvenus) [inline examples/example_push.rs]: > we must tell user to start prometheus server first.

It should be [pushgateway](https://github.com/prometheus/pushgateway), not prometheues server.

> Add a flag for the prometheus server, not fixed 127.0.0.1:9091

`127.0.0.1:9091` is the default address of pushgateway

-------------

I will address comments later.

File: src/push.rs [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    let max_len = 256;
+    let mut name = vec![0u8; max_len];
+    match unsafe {
+        gethostname(name.as_mut_ptr() as *mut libc::c_char,
+                    max_len as libc::size_t)
+    } {
+        0 => {
+            let last_char = name.iter().position(|byte| *byte == 0).unwrap_or(max_len);
+            labels!{
+                "instance".to_owned() => str::from_utf8(&name[..last_char])
```
Comment (siddontang) [inline src/push.rs]: define const vars for "instance" and "unknown".

File: examples/example_push.rs (hunk 7 above)
Comment (hhkbp2) [inline examples/example_push.rs]: A super computer in the movie 2001: A Space Odyssey.

File: src/push.rs [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        url_components.push(lv.to_owned());
+    }
+
+    push_url = format!("{}/metrics/job/{}", push_url, url_components.join("/"));
+
+    let encoder = TextEncoder::new();
+    let mut buf = Vec::new();
+    try!(encoder.encode(&mfs, &mut buf));
+
+    let request = HTTP_CLIENT.request(Method::from_str(method).unwrap(), &push_url)
```
Comment (disksing) [inline src/push.rs]: No need to check for pre-existing grouping labels?
https://github.com/prometheus/client_golang/blob/master/prometheus/push/push.go#L103

File: examples/example_push.rs [hunk 10]
```diff
@@ -38,16 +42,33 @@ lazy_static! {
...
 
 fn main() {
-    println!("Pushing, please wait 10 seconds ...");
+    let args: Vec<String> = env::args().collect();
+    let program = args[0].clone();
+
+    let mut opts = Options::new();
+    opts.optopt("A",
+                "addr",
+                "set pushgateway address",
```
Comment (siddontang) [inline examples/example_push.rs]: /set/prometheus/

**Comment by disksing** [comment]
LGTM

File: src/push.rs [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+        url_components.push(lv.to_owned());
+    }
+
+    push_url = format!("{}/metrics/job/{}", push_url, url_components.join("/"));
+
+    // Check for pre-existing grouping labels:
+    for mf in &mfs {
+        for m in mf.get_metric() {
+            for lp in m.get_label() {
+                if lp.get_name() == "job" {
```
Comment (siddontang) [inline src/push.rs]: define a const var

Comment (overvenus) [inline src/push.rs]: I think it's more readable.

Comment (siddontang) [inline src/push.rs]: no, define it.

File: src/push.rs [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+/// provided collectors. It is a convenient way to push only a few metrics.
+pub fn push_add_collector(job: &str,
+                          grouping: HashMap<String, String>,
+                          url: &str,
+                          collectors: Vec<Box<Collector>>)
+                          -> Result<()> {
+    push_from_collector(job, grouping, url, collectors, "POST")
+}
+
+const DEFAULT_GROUP_LABEL_PAIR: (&'static str, &'static str) = ("intance", "unknown");
```
Comment (overvenus) [inline src/push.rs]: Oops, typo.

File: src/push.rs (hunk 11 above)
Comment (siddontang) [inline src/push.rs]: we should reject any magic number, string, except very frequent use like "GET", "POST", etc.

Comment (overvenus) [inline src/push.rs]: Got it.

Comment (siddontang) [inline src/push.rs]: any test can cover this?

File: src/push.rs [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    let max_len = 256;
+    let mut name = vec![0u8; max_len];
+    match unsafe {
+        gethostname(name.as_mut_ptr() as *mut libc::c_char,
+                    max_len as libc::size_t)
+    } {
+        0 => {
+            let last_char = name.iter().position(|byte| *byte == 0).unwrap_or(max_len);
+            labels!{
+                DEFAULT_GROUP_LABEL_PAIR.0.to_owned() => str::from_utf8(&name[..last_char])
```
Comment (siddontang) [inline src/push.rs]: seem that the key is "instant" , why not using two const vars? , no need tuple.

File: src/push.rs [hunk 14]
```diff
@@ -111,7 +114,7 @@ fn push(job: &str,
     for mf in &mfs {
         for m in mf.get_metric() {
             for lp in m.get_label() {
-                if lp.get_name() == "job" {
+                if lp.get_name() == INVALID_PUSH_LABEL_NAME_JOB {
                     return Err(Error::Msg(format!("pushed metric {} already contains a \
```
Comment (siddontang) [inline src/push.rs]: pushed metric can't use job label.

File: src/push.rs (hunk 13 above)
Comment (overvenus) [inline src/push.rs]: So we know they are paired.

File: src/push.rs [hunk 15]
```diff
@@ -71,6 +71,9 @@ pub fn push_add_metrics(job: &str,
     push(job, grouping, url, mfs, "POST")
 }
 
+// pub for tests
+pub const INVALID_PUSH_LABEL_NAME_JOB: &'static str = "job";
```
Comment (siddontang) [inline src/push.rs]: PUSH_JOB_LABEL_NAME

**Comment by siddontang** [comment]
Rest LGTM

File: src/push.rs (hunk 14 above)
Comment (overvenus) [inline src/push.rs]: Just copied from go, I think we'd better using client_golang error message.

Comment (siddontang) [inline src/push.rs]: got it

**Pull request merged**